sylt/src/formatter.rs

```diff
@@ -0,0 +1,504 @@
+use std::fmt::{self, Write};
+use std::path::PathBuf;
+use sylt_common::{Error, Type as RuntimeType};
+use sylt_parser::statement::UseIdentifier;
+use sylt_parser::{
+    Assignable, AssignableKind, Expression, ExpressionKind, Identifier, Module, Op, Statement,
+    StatementKind, Type, TypeKind, VarKind,
+};
+
+use crate::Args;
+
+static INDENT: &'static str = "    ";
+
+macro_rules! write_comma_separated {
+    ($dest:expr, $indent:expr, $write:expr, $values:expr) => {
+        let mut first = true;
+        for value in $values {
+            if !first {
+                write!($dest, ", ")?;
+            }
+            first = false;
+            $write($dest, $indent, value)?;
+        }
+    };
+}
+
+fn write_indents<W: Write>(dest: &mut W, indent: u32) -> fmt::Result {
+    for _ in 0..indent {
+        write!(dest, "{}", INDENT)?;
+    }
+    Ok(())
+}
+
+fn write_identifier<W: Write>(dest: &mut W, identifier: &Identifier) -> fmt::Result {
+    write!(dest, "{}", identifier.name)
+}
+
+fn write_parameters<W: Write>(
+    dest: &mut W,
+    indent: u32,
+    parameters: &[(Identifier, Type)],
+) -> fmt::Result {
+    let mut first = true;
+    for (identifier, ty) in parameters {
+        if !first {
+            write!(dest, ", ")?;
+        }
+        first = false;
+        write_identifier(dest, identifier)?;
+        write!(dest, ": ")?;
+        write_type(dest, indent, ty)?;
+    }
+    Ok(())
+}
+
+fn write_blob_instance_fields<W: Write>(
+    dest: &mut W,
+    indent: u32,
+    fields: &[(String, Expression)],
+) -> fmt::Result {
+    for (field, expr) in fields {
+        write_indents(dest, indent)?;
+        write!(dest, "{}: ", field)?;
+        write_expression(dest, indent, expr)?;
+        write!(dest, "\n")?;
+    }
+    Ok(())
+}
+
+fn write_runtime_type<W: Write>(dest: &mut W, ty: &RuntimeType) -> fmt::Result {
+    match ty {
+        RuntimeType::Void => write!(dest, "nil"),
+        RuntimeType::Int => write!(dest, "int"),
+        RuntimeType::Float => write!(dest, "float"),
+        RuntimeType::Bool => write!(dest, "bool"),
+        RuntimeType::String => write!(dest, "str"),
+        t => unreachable!("{:?} should not be present in the syntax tree", t),
+    }
+}
```

Comment:
This could probably be replaced now. I would double check that the Void type looks correct though.

Reply:
double :+1: